### ISSUE TITLE
Debugging InvalidSystemClockExceptions

### ISSUE THREAD

**Issue by Mahamed-Belkheir** [RobThree/IdGen#59]
We're running into sporadic InvalidSystemClockExceptions that end up requiring restarts, and we're running out of ideas trying to find where it goes wrong.

one error example is:
`IdGen.InvalidSystemClockException: Clock moved backwards or wrapped around. Refusing to generate id for 4294446821 ticks`
from reading the code, this implies the new timestamp is behind the last generated one by 50 days, which seems too big of a difference even if there was a clock sync issue (we've checked and our k8s nodes are running chrony and the recorded sync errors are nowhere near that much of a difference).

The issue seems to happen once a month or so, and forces us to manually restart pods as it blocks all writes, and waiting the ticks for it to restart would take days.

Our timesource implementation
```c#
public class ClockTimeSource : ITimeSource
{
    private long previousTimestamp = 0;

    public DateTimeOffset Epoch { get; }
    public TimeSpan TickDuration => TimeSpan.FromMilliseconds(1);

    public ClockTimeSource() { }
    public ClockTimeSource(DateTimeOffset epoch) => Epoch = epoch;

    public long GetTicks()
    {
        var timestamp = (long)(DateTimeOffset.UtcNow - Epoch).TotalMilliseconds;
        if (timestamp > previousTimestamp)
        {
            previousTimestamp = timestamp;
        }
        return previousTimestamp;
    }
}
```

Would appreciate any pointers on where to start debugging this.

**Comment by RobThree** [comment]
> Would appreciate any pointers on where to start debugging this.

Well, you're using `DateTimeOffset.UtcNow` so the issue must lie somewhere in there. If I had to debug this issue I'd start by logging the actual vallue of `DateTimeOffset.UtcNow` at the time of the exception. It should at least show / demonstrate that the value is, indeed, ~~50~~ 49.7 days off. If it isn't then the problem must lie elsewhere.

However, `4294446821` is *awfully* close to the max value of an unsigned 32 bit int (`4294967295`) (...and that takes me back to the time where [Windows would crash after 49.7 days](https://www.cnet.com/culture/windows-may-crash-after-49-7-days/) 🤣 ). That would imply an `int` is used somewhere. That, in turn, would mean that `DateTimeOffset.UtcNow - Epoch` somehow uses an int somewhere - which I don't believe it does. But I *am* wondering what your epoch is. Since you mention that restarting fixes the problem for a month, your epoch should be 'moving' a month or so somehow?

A [quick test](https://dotnetfiddle.net/2Mr9D1) seems to confirm that a long is returned correctly... So I suspect you're casting this value to an `int` somewhere?

**Comment by Mahamed-Belkheir** [comment]
> Since you mention that restarting fixes the problem for a month, your epoch should be 'moving' a month or so somehow?

I also thought that should be the case, and restarting should not be fixing the issue, since our epoch is defined as a constant `public const long Epoch = 1666479867;`
and passed in as
```c#
() => new IdGeneratorOptions
{
    IdStructure = new IdStructure(32, 16, 15),
    TimeSource = new ClockTimeSource(DateTimeOffset.FromUnixTimeSeconds(Epoch)),
    SequenceOverflowStrategy = SequenceOverflowStrategy.SpinWait,
});
```

It completely passed me that value is very close to the 32uint max value, if an int is overflowing and going back to 0 it also explains the frequency of the issue happening, so that's a good lead to follow, thank you 🙏

**Comment by RobThree** [comment]
If you find out the problem, please let me know? I'm curious to what it was.

Also out of curiosity: your epoch is `2022-10-22 23:04:27 +00:00`. This is intentional? (Just checking because it's a bit of a weird date/time to be an epoch...).

**Comment by hugh-maaskant** [comment]
Hi, just chiming in here ...

If `GetTicks()` is called from multiple threads, I believe there also are two problems in its implementation:

1. The check `timestamp > previousTimestamp` and the assignment of the new timestamp value need to be atomic. In the really unlikely case that the thread checking is scheduled out just after the check and then it takes a long time to get rescheduled (more than the tick time - so as I said unlikely) than it could reassign a lower `timestamp` value back to the `previousTimestamp` that was meanwhile update twice by other threads
2. The assignment of a long value itself is not atomic on all processor architectures, so (again unlikely) it could get the upper half of one value and the lower half of another value (from another thread).

Both of these are easily (and fairly cheaply) avoided by using a `lock()` statement: [https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock).

Good luck :-)

**Comment by Mahamed-Belkheir** [comment]
Will leave the issue open while investigating, also curiously enough just noticed the IdStructure takes 32 bits for the timestamp part, another coincidence? going to test if the time mask rounds out some timestamp values to 0

> Also out of curiosity: your epoch is 2022-10-22 23:04:27 +00:00. This is intentional? (Just checking because it's a bit of a weird date/time to be an epoch...).

I'm not the author of the code, rather a recent onboarding, so I'm not sure, but I'll ask

**Comment by RobThree** [comment]
> Hi, just chiming in here ...
> 
> If `GetTicks()` is called from multiple threads, I believe there also are two problems in its implementation:
> 
> 1. The check `timestamp > previousTimestamp` and the assignment of the new timestamp value need to be atomic. In the really unlikely case that the thread checking is scheduled out just after the check and then it takes a long time to get rescheduled (more than the tick time - so as I said unlikely) than it could reassign a lower `timestamp` value back to the `previousTimestamp` that was meanwhile update twice by other threads
> 2. The assignment of a long value itself is not atomic on all processor architectures, so (again unlikely) it could get the upper half of one value and the lower half of another value (from another thread).
> 
> Both of these are easily (and fairly cheaply) avoided by using a `lock()` statement: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock.
> 
> Good luck :-)

Ooooh, wait. I think you are correct. And the value being a long may result in a torn-read! Ofcourse! Hmmm, will investigate.

Wait, never mind 🤦 The code for `timestamp > previousTimestamp` is not in IdGen, and [where it is](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdGenerator.cs#L120), it is in a lock 😄 
And this would explain the 49.7 days, after that the value because bigger than 32 bits and THUS 2 writes, resulting in a torn read in high load situations.

**Comment by Mahamed-Belkheir** [comment]
> If GetTicks() is called from multiple threads

I had to double check, and it is indeed defined as a singleton, so it's likely running into multithreading issues, will update the time source code, but this is a problem that happens every month or so I'll be closing the issue, will re-open if it still occurs.

Thanks for the help!

**Comment by RobThree** [comment]
> I had to double check, and it is indeed defined as a singleton

It being a singleton is exactly the problem; multiple threads are sharing the object and setting the `previousTimestamp` While one thread is setting the value (writing 32 bits and then another 32 bits) another thread does a read on the first 32 bits 'in-between' the other threads writes. This results in a torn read and, as I said here, results in exactly the behaviour you see. As long as you're under the 49.7 days you're safe because write of 32 bits is atomic, 64 bits isn't. After 49.7 days you 'overflow' into the second 32 bit "half" of the value and _thus_ from that point on you're susceptible to a torn read. I'm 99.9% sure this is your problem.

This may explain it better: https://benbowen.blog/post/cmmics_iii/

**Comment by Mahamed-Belkheir** [comment]
One thing though, it does seem like `GetTicks` only gets called inside of the lock section of `CreateIdImpl`, I'm not well versed in the .net world, but shouldn't this also keep the time source `GetTicks` method synced up?

**Comment by RobThree** [comment]
Yes, but is IdGen the only thing accessing your `ClockTimeSource`? Or are you using the `ClockTimeSource` for other purposes as well?

@hugh-maaskant great spot! 👍

**Comment by Mahamed-Belkheir** [comment]
> Yes, but is IdGen the only thing accessing your `ClockTimeSource`? Or are you using the `ClockTimeSource` for other purposes as well?
> 
> @hugh-maaskant great spot! 👍

It is passed as a new instance to IdGeneratorOptions as shared earlier (more context added)
```c#
public static IServiceCollection AddSnowflakeGenerator(this IServiceCollection collection, string ipAddress)
{
    if (ipAddress == null)
    {
        throw new ArgumentNullException(nameof(ipAddress));
    }

    return collection.AddIdGen( 
        ConvertFromIpAddressToInteger(ipAddress),
        () => new IdGeneratorOptions
        {
            IdStructure = new IdStructure(32, 16, 15),
            TimeSource = new ClockTimeSource(DateTimeOffset.FromUnixTimeSeconds(MikiEpoch)),
            SequenceOverflowStrategy = SequenceOverflowStrategy.SpinWait,
        });
}
```
ClockTimeSource itself isn't the singleton, IdGen is.


But going over the code, I've noticed that our IdStructure is using 32 bits for the timestamp, which after going over the code for IdGen, it is probably the cause, as it gets bitwise AND with the mask and starts wrap over after the max uint32 value `4294967296` https://dotnetfiddle.net/auP2Zb

This also explains why it happens periodically and a restart fixes it, it might start at tick 0 or 4294967295, or any other value in between, but only fails once it lives long enough to wrap again once more and have a lesser value than the last tick.

**Comment by RobThree** [comment]
> ClockTimeSource itself isn't the singleton, IdGen is.

But what is the reason you have a custom [`ClockTimeSource`](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdGenerator.cs#L170) in the first place? I'm still kind of convinced you're using it outside of IdGen too - which doesn't have to be a problem as long as proper locking or other concurrency control is used.

> But going over the code, I've noticed that our IdStructure is using 32 bits for the timestamp, which after going over the code for IdGen, it is probably the cause, as it gets bitwise AND with the mask and starts wrap over after the max uint32 value `4294967296` https://dotnetfiddle.net/auP2Zb

Then wouldn't you get a compiler warning because you're casting a long to an int? And, again, purely out of curiosity, why is the Id used for some calculation at all? It is supposed to be used as an opaque identifier and that's about it. You shouldn't do anything with it but generate it and use it as an opaque id. The only - sort of - thing you can do with it is pass it to [`FromId`](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdGenerator.cs#L170) to get the timestamp etc. back but I wouldn't even recommend that.

Would you mind sharing with us why you're doing calculations on the generated id and/or maybe show the relevant code?

> This also explains why it happens periodically and a restart fixes it, it might start at tick 0 or 4294967295, or any other value in between

Then it wouldn't explain it crashing repeatably at 49.7 days; then it would crash anywhere between 0 and 49.7 days.

> but only fails once it lives long enough to wrap again once more and have a lesser value than the last tick

As said, I'm still convinced the problem is a torn read in your `ClockTimeSource` and the simples solution would be to add a `lock` in the `GetTicks` method. Your reasoning also doesn't explain why you get a `InvalidSystemClockException`, which is thrown _during_ generation of an Id. If the problem was with you doing calculations involving an int and something overflowing, there wouldn't be a `InvalidSystemClockException` because you're doing the calculations on a generated ("completed") Id.

**Comment by RobThree** [comment]
Uh, wait... I just spotted [this](https://github.com/RobThree/IdGen/issues/59#issuecomment-1759550634):

```c#
IdStructure = new IdStructure(32, 16, 15)
```

Why are you reserving only 32 bits for the timestamp part? That is at the very least _also_ a problem, if not _the_ problem. Just because IdGen lets you, it doesn't mean it's a good idea 😅 

If you insist on a 32 bit timestamp part then your timesource "resolution" should drop to _at least_ a second. That will still give you less than 100 years of 'range' which, depending on your epoch, may or may not be a [problem](https://en.wikipedia.org/wiki/Year_2038_problem). Since your epoch is in 2022 - and assuming you're going from milliseconds to seconds - you'll overflow somewhere in 2090 or so. But if you keep using milliseconds and only 32 bits for the timestamp part then you'll overflow every 49.7 days - guaranteed.

I should've spotted this earlier, don't know why I overlooked it. But this doesn't make sense. The problem is still exactly the same though; it's at the 32-bit boundary. Wether it overflows or is a torn read and a concurrency issue isn't (so much) relevant. The IdStructure chosen (32 bits for the timestamp) is just a poor decision. Do you really need 16 bits (64K) for the generator part? Maybe you can explain the reasoning behind the chosen IdStructure so we can help advise on that matter? Also: If you're already in production with these settings then you may be in for some [pain](https://github.com/RobThree/IdGen#faq):

> A: Also: Don't change the structure; once you've picked an `IdStructure` and go into production commit to it, stick with it. This means that careful planning is needed to ensure enough ID's can be generated by enough generators for long enough. Although changing the structure at a later stage isn't impossible, careful consideration is needed to ensure no collisions will occur.

Long story short: Either increase the tick duration to something like a second or reserve more bits for the timestamp part. You can use the [`WraparoundDate()`](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdStructure.cs#L89) and/or [`WraparoundInterval()`](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdStructure.cs#L113) to calculate when, or at what interval, your chosen IdStructure will overflow and - thus - become a problem.

**Comment by Mahamed-Belkheir** [comment]
> Then wouldn't you get a compiler warning because you're casting a long to an int? And, again, purely out of curiosity, why is the Id used for some calculation at all? It is supposed to be used as an opaque identifier and that's about it. You shouldn't do anything with it but generate it and use it as an opaque id. The only - sort of - thing you can do with it is pass it to [FromId](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdGenerator.cs#L170) to get the timestamp etc. back but I wouldn't even recommend that.

The part you replied to was in reference to how IdGen works, not our own code, i meant [this part](https://github.com/RobThree/IdGen/blob/ce5e341e34a51a39c9dda47e9cd299a8339aea7d/IdGen/IdGenerator.cs#L118C24-L118C24) which is working as intended

Also, at this point I'm almost certain our ID structure is the main/only issue, ClockTimeSource isn't used elsewhere.

> Then it wouldn't explain it crashing repeatably at 49.7 days; then it would crash anywhere between 0 and 49.7 days.

it ~~crashes~~ starts failing randomly, the 49.7 days is the clock offset we get in the exception (reaches max uint32 value then wraps over), it's simply guaranteed to get stuck before 49.7 days, but can do so before it.

>  Do you really need 16 bits (64K) for the generator part? Maybe you can explain the reasoning behind the chosen IdStructure so we can help advise on that matter?

We're building the generator id based on the pod ip (e.g `x.x.node_ip.pod_ip`) which is guaranteed to be unique for each pod, the other alternatives we thought of are using stateful sets or building an gen ID lending service.

We're currently thinking of switching to `41, 14, 8`, we figured we do not need as much IDs per pod per second as originally set, and we have  shrunk the section for node_ips from 8 bits/256 values to 6 bites, so around 60 nodes maximum as we can safely assume we won't be scaling above that before 41 bits timestamp wraps around anyway.

for our usecase, ID collisions are not the end of the world, and services blocking writes are a bigger issue, though if you have some migration suggestions it would be appreciated, we thought of regenerating all the IDs, but we've got around 100m records so preferably not, we've also considered going over the 64bit limit for new ids (db schema already supports that) but that would require a fork of IdGen or patching of all the IdGen using code snippets

**Comment by Mahamed-Belkheir** [comment]
edit: accidental double comment
